plugin/kubernetes/controller.go

```diff
@@ -341,62 +218,59 @@ func (dns *dnsControl) Stop() error {
 // Run starts the controller.
 func (dns *dnsControl) Run() {
 	go dns.svcController.Run(dns.stopCh)
-	go dns.nsController.Run(dns.stopCh)
 	go dns.epController.Run(dns.stopCh)
 	if dns.podController != nil {
 		go dns.podController.Run(dns.stopCh)
 	}
 	<-dns.stopCh
 }
 
-func (dns *dnsControl) NamespaceList() *api.NamespaceList {
-	nsList, err := dns.nsLister.List()
-	if err != nil {
-		return &api.NamespaceList{}
-	}
-
-	return &nsList
-}
-
-func (dns *dnsControl) ServiceList() []*api.Service {
-	svcs, err := dns.svcLister.List(labels.Everything())
-	if err != nil {
-		return []*api.Service{}
+func (dns *dnsControl) ServiceList() (svcs []*api.Service) {
+	os := dns.svcLister.List()
+	for _, o := range os {
+		s, ok := o.(*api.Service)
+		if !ok {
+			continue
+		}
+		svcs = append(svcs, s)
 	}
-
 	return svcs
 }
 
-func (dns *dnsControl) PodIndex(ip string) []interface{} {
-	if dns.podLister.Indexer == nil {
+func (dns *dnsControl) PodIndex(ip string) (pods []*api.Pod) {
+	if dns.podLister == nil {
 		return nil
 	}
-	pods, err := dns.podLister.Indexer.ByIndex(podIPIndex, ip)
+	os, err := dns.podLister.ByIndex(podIPIndex, ip)
 	if err != nil {
 		return nil
 	}
-
+	for _, o := range os {
+		p, ok := o.(*api.Pod)
```

Comment:
again, is this really needed? Or can this API really return something else?

Reply:
I dont know. I was coding defensively...  Should I remove the checks?

Reply:
dunno... @johnbelamaric any ideas?

Reply:
we can remove them later.

Reply:
> It's an extra loop for a k8s queries..

The loop is required to convert the types to the return type.  This is why I was doing the conversion outside the function originally.

> Are these checks really needed

This i dont know for sure ... But i think that if they always returned one type, the methods would be typed that way, hence the check.

Reply:
I wonder how the well type assertions are optimized by the compiler. I would imagine the overhead would be really small.

Reply:
Also none of this stuff is in-band.

Reply:
I mean, none of this runs for a DNS query, it's all async stuff in other go routines.